File: frontend/src/services/reports.ts
import { apiClient } from "./apiClient";

export interface ReportSummary {
  expeditionReadinessPct: number;
  cargoTonnageTracked: number;
  criticalSupplyDaysMin: number;
  totalMissionsCompleted: number;
  activeIncidentsCount: number;
  fuelEfficiencyRate: string;
}

export interface MonthlyCargoChartItem {
  month: string;
  dispatched: number;
  received: number;
}

export interface FuelConsumptionChartItem {
  week: string;
  maitri: number;
  bharati: number;
}

export interface AssetHealthChartItem {
  name: string;
  score: number;
  total: number;
  operational: number;
}

export interface ReportChartsResponse {
  monthlyCargoData: MonthlyCargoChartItem[];
  fuelConsumptionData: FuelConsumptionChartItem[];
  assetHealthData: AssetHealthChartItem[];
  totalDeliveredTonnes: number;
}

export const reportsService = {
  async getPerformanceSummary(): Promise<ReportSummary> {
    try {
      const data = await apiClient.get<any>("/reports/summary");
      return {
        expeditionReadinessPct:
          data.expeditionReadinessPct ?? data.expedition_readiness_pct ?? 0,
        cargoTonnageTracked:
          data.cargoTonnageTracked ?? data.cargo_tonnage_tracked ?? 0,
        criticalSupplyDaysMin:
          data.criticalSupplyDaysMin ?? data.critical_supply_days_min ?? 0,
        totalMissionsCompleted:
          data.totalMissionsCompleted ?? data.total_missions_completed ?? 0,
        activeIncidentsCount:
          data.activeIncidentsCount ?? data.active_incidents_count ?? 0,
        fuelEfficiencyRate:
          data.fuelEfficiencyRate ?? data.fuel_efficiency_rate ?? "N/A",
      };
    } catch (err: any) {
      if (err?.isOffline) {
        return {
          expeditionReadinessPct: 0,
          cargoTonnageTracked: 0,
          criticalSupplyDaysMin: 0,
          totalMissionsCompleted: 0,
          activeIncidentsCount: 0,
          fuelEfficiencyRate: "N/A (Offline)",
        };
      }
      throw err;
    }
  },

  async getReportCharts(): Promise<ReportChartsResponse> {
    try {
      const data = await apiClient.get<any>("/reports/charts");
      return {
        monthlyCargoData: data.monthlyCargoData ?? [],
        fuelConsumptionData: data.fuelConsumptionData ?? [],
        assetHealthData: data.assetHealthData ?? [],
        totalDeliveredTonnes: data.totalDeliveredTonnes ?? 0,
      };
    } catch (err: any) {
      if (err?.isOffline) {
        return {
          monthlyCargoData: [],
          fuelConsumptionData: [],
          assetHealthData: [],
          totalDeliveredTonnes: 0,
        };
      }
      throw err;
    }
  },
};

File: backend/app/schemas/reports.py
from pydantic import BaseModel, ConfigDict, Field
from typing import List, Optional


class ReportSummaryResponse(BaseModel):
    expedition_readiness_pct: float = Field(..., validation_alias="expeditionReadinessPct", serialization_alias="expeditionReadinessPct")
    cargo_tonnage_tracked: float = Field(..., validation_alias="cargoTonnageTracked", serialization_alias="cargoTonnageTracked")
    critical_supply_days_min: float = Field(..., validation_alias="criticalSupplyDaysMin", serialization_alias="criticalSupplyDaysMin")
    total_missions_completed: int = Field(..., validation_alias="totalMissionsCompleted", serialization_alias="totalMissionsCompleted")
    active_incidents_count: int = Field(..., validation_alias="activeIncidentsCount", serialization_alias="activeIncidentsCount")
    fuel_efficiency_rate: str = Field(..., validation_alias="fuelEfficiencyRate", serialization_alias="fuelEfficiencyRate")

    model_config = ConfigDict(
        populate_by_name=True,
        from_attributes=True,
    )


class MonthlyCargoChartItem(BaseModel):
    month: str
    dispatched: float
    received: float


class FuelConsumptionChartItem(BaseModel):
    week: str
    maitri: float
    bharati: float


class AssetHealthChartItem(BaseModel):
    name: str
    score: float
    total: int
    operational: int


class ReportChartsResponse(BaseModel):
    monthly_cargo_data: List[MonthlyCargoChartItem] = Field(..., validation_alias="monthlyCargoData", serialization_alias="monthlyCargoData")
    fuel_consumption_data: List[FuelConsumptionChartItem] = Field(..., validation_alias="fuelConsumptionData", serialization_alias="fuelConsumptionData")
    asset_health_data: List[AssetHealthChartItem] = Field(..., validation_alias="assetHealthData", serialization_alias="assetHealthData")
    total_delivered_tonnes: float = Field(..., validation_alias="totalDeliveredTonnes", serialization_alias="totalDeliveredTonnes")

    model_config = ConfigDict(
        populate_by_name=True,
        from_attributes=True,
    )
